core/hotspot/slot.go

```diff
@@ -41,57 +40,35 @@ func (s *Slot) Order() uint32 {
 	return RuleCheckSlotOrder
 }
 
-// matchArg matches the arg from args based on TrafficShapingController
-// return nil if match failed.
-func matchArg(tc TrafficShapingController, args []interface{}) interface{} {
-	if tc == nil {
-		return nil
-	}
-	idx := tc.BoundParamIndex()
-	if idx < 0 {
-		idx = len(args) + idx
-	}
-	if idx < 0 {
-		if logging.DebugEnabled() {
-			logging.Debug("[Slot matchArg] The param index of hotspot traffic shaping controller is invalid", "args", args, "paramIndex", tc.BoundParamIndex())
-		}
-		return nil
-	}
-	if idx >= len(args) {
-		if logging.DebugEnabled() {
-			logging.Debug("[Slot matchArg] The argument in index doesn't exist", "args", args, "paramIndex", tc.BoundParamIndex())
-		}
-		return nil
-	}
-	return args[idx]
-}
-
 func (s *Slot) Check(ctx *base.EntryContext) *base.TokenResult {
 	res := ctx.Resource.Name()
-	args := ctx.Input.Args
+
 	batch := int64(ctx.Input.BatchCount)
 
 	result := ctx.RuleCheckResult
 	tcs := getTrafficControllersFor(res)
 	for _, tc := range tcs {
-		arg := matchArg(tc, args)
-		if arg == nil {
-			continue
-		}
-		r := canPassCheck(tc, arg, batch)
-		if r == nil {
+		args := tc.ExtractArgs(ctx)
+		if args == nil || len(args) == 0 {
 			continue
 		}
-		if r.Status() == base.ResultStatusBlocked {
-			return r
-		}
-		if r.Status() == base.ResultStatusShouldWait {
-			if nanosToWait := r.NanosToWait(); nanosToWait > 0 {
-				// Handle waiting action.
-				time.Sleep(nanosToWait)
+		for _, arg := range args {
```

Comment:
larger len(args) , the lower the efficiency will be, concurrent processing should be considered.

Reply:
![image](https://user-images.githubusercontent.com/13758306/102855449-a5d95e00-445f-11eb-9937-82cb1114e66e.png)